### PR TITLE
Add new gitlab-runner-scaler

### LINKED ISSUE #5616: GitLab Runner Scaler
### Proposal

KEDA already supports GitHub Runners. As a self-hosted GitLab user, it'd be great to have support for GitLab Runners.

### Scaler Source

https://docs.gitlab.com/runner/

### Scaling Mechanics

https://docs.gitlab.com/runner/fleet_scaling/

### Authentication Source

https://docs.gitlab.com/ee/ci/runners/configure_runners.html#authentication-token-security

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by fira42073**
_Provide a description of what has been changed_

### Checklist

- [x] When introducing a new scaler, I agree with the [scaling governance policy](https://github.com/kedacore/governance/blob/main/SCALERS.md)
- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Tests have been added
- [ ] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [ ] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [ ] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [ ] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

Fixes #5616

### CODE REVIEW

**Comment by fira42073** [comment]
https://docs.gitlab.com/ee/api/pipelines.html#list-project-pipelines

File: pkg/scalers/gitlab_runner_scaler.go [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+
+	return &gitlabRunnerScaler{
+		metricType: metricType,
+		metadata:   meta,
+		httpClient: httpClient,
+		logger:     InitializeLogger(config, "gitlab_runner_scaler"),
+	}, nil
+}
+
+func parseGitLabRunnerMetadata(config *scalersconfig.ScalerConfig) (*gitlabRunnerMetadata, error) {
```
Comment (zroubalik) [inline pkg/scalers/gitlab_runner_scaler.go:72]: Could you please use the new declarative approach for the metadata parsing?

https://github.com/kedacore/keda/issues/5797

File: pkg/scalers/gitlab_runner_scaler.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+	v2 "k8s.io/api/autoscaling/v2"
+	"k8s.io/metrics/pkg/apis/external_metrics"
+
+	"github.com/kedacore/keda/v2/pkg/scalers/scalersconfig"
+	kedautil "github.com/kedacore/keda/v2/pkg/util"
+)
+
+const (
+	// externalMetricType is the type of the external metric.
+	defaultTargetPipelineQueueLength = 1
```
Comment (zroubalik) [inline pkg/scalers/gitlab_runner_scaler.go:21]: this comment is not relevant

File: pkg/scalers/gitlab_runner_scaler.go (hunk 1 above)
Comment (fira42073) [reply pkg/scalers/gitlab_runner_scaler.go:72]: this declarative approach is so much more convenient! thanks for pointing it out, I'll fix it.

File: pkg/scalers/scalersconfig/typed_config.go [hunk 3]
```diff
@@ -273,6 +273,22 @@ func setConfigValueURLParams(params Params, valFromConfig string, field reflect.
...
+	}
+
+	// If the field type is a pointer to url.URL (`*url.URL`), set the value directly
+	if field.Kind() == reflect.Ptr && field.Type().Elem() == reflect.TypeOf(url.URL{}) {
+		field.Set(reflect.ValueOf(u))
+		return nil
+	}
+
+	return nil
+}
```
Comment (wozniakjan) [inline pkg/scalers/scalersconfig/typed_config.go:290]: can you please add a simple unit test for this? you can take an inspiration in
https://github.com/kedacore/keda/blob/5b7a372bf249ad6a2a280ee71efc1677b098b04e/pkg/scalers/scalersconfig/typed_config_test.go#L27

File: pkg/scalers/scalersconfig/typed_config.go [hunk 4]
```diff
@@ -273,6 +273,22 @@ func setConfigValueURLParams(params Params, valFromConfig string, field reflect.
 	return nil
 }
 
+// setConfigValueURL is a function that sets the value of the url.URL field
+func setConfigValueURL(_ Params, valFromConfig string, field reflect.Value) error {
```
Comment (wozniakjan) [inline pkg/scalers/scalersconfig/typed_config.go:277]: per our discussion in https://github.com/kedacore/keda/issues/5616#issuecomment-2310567282, you don't have to pass arguments that the function body doesn't need
```suggestion
func setConfigValueURL(valFromConfig string, field reflect.Value) error {
```

File: pkg/scalers/scalersconfig/typed_config.go (hunk 3 above)
Comment (fira42073) [reply pkg/scalers/scalersconfig/typed_config.go]: Done

**Pull request closed without merging**